src/GridFieldActionsMenu.php

```diff
@@ -0,0 +1,277 @@
+            /** @var \SilverStripe\Forms\Tab $tab */
+            $tabID = ($first) ? null : $tab->ID();
+            $link = Controller::join_links($gridField->Link('item'), $record->ID, 'edit');
+            // @TODO hack workaround: `&& false` here because some JS in the CMS is rewriting
+            // a link with a hash in it to the page we're on _now_ #anchor, as opposed to
+            // e.g link/set/here#anchor
+            $link = $tabID && false ? "$link#$tabID" : $link;
+
+            $actions[] = [
+                'Title' => $tab->Title(),
+                'Link' => $link,
+                'Type' => $typeName
+            ];
+            $first = false;
+        }
+        return $actions;
+    }
+
+    /**
+     * If the object is versioned (has the {@link Versioned} extension applied) then add
+     * actions to publish, unpublish, and archive - depending on record state.
+     *
+     * @param DataObject $record
+     * @param string $typeName
+     *
+     * @return array
+     */
+    protected function addVersionedActions(GridField $gridField, DataObject $record, $typeName)
+    {
+        if (!$record->hasExtension(Versioned::class)) {
+            return [];
+        }
+
+        $actions = [];
+
+        if (!$record->latestPublished()) {
+            $actions[] = [
+                'Title' => _t(__CLASS__ . '.Publish', 'Publish'),
+                'Link' => Controller::join_links($gridField->Link('item'), $record->ID, 'publish'),
+                'Type' => $typeName
+            ];
+        }
+
+        if ($record->isPublished()) {
+            $actions[] = [
+                'Title' => _t(__CLASS__ . '.Unpublish', 'Unpublish'),
+                'Link' => Controller::join_links($gridField->Link('item'), $record->ID, 'unpublish'),
+                'Type' => $typeName
+            ];
+        }
+
+        $actions[] = [
+            'Title' => _t(__CLASS__ . '.Delete', 'Delete'),
+            'Link' => Controller::join_links($gridField->Link('item'), $record->ID, 'archive'),
+            'Type' => $typeName
+        ];
+
+        return $actions;
+    }
+
+    public function getColumnContent($gridField, $record, $columnName)
+    {
+        $actionBuilder = function ($item, $section) use ($gridField, $record) {
+            $type = is_string($section) ? $section : null;
+            $result = (is_callable($item)) ?
+                $item($gridField, $record, $type) :
+                $item;
+            return $result;
+        };
+        // compile our action lists
+        $actions = $this->getActions();
+        $actions = array_map($actionBuilder, $actions, array_keys($actions));
+        if (!$this->getShowFirstTab()) {
+            reset($actions);
+            $first = key($actions);
+            array_shift($actions[$first]);
+        }
+        // remove empty entries
+        foreach ($actions as $section => $sectionItems) {
+            $section = array_filter($sectionItems);
```

Comment:
This variable is never used for anything; My IDE detects it as unused. Should something happen here?

Reply:
Hmm, there's supposed to be a deeper dereferencing there, indeed!